### ISSUE TITLE
\text_titlecase:n and string variables

### ISSUE THREAD

**Issue by wehro** [latex3/latex3#984]
The `\text_titlecase:n` command does not seem to work correctly with string variables. The following document gives me the output “ABC”, but it should be “Abc”.
```
\documentclass{article}
\begin{document}
\ExplSyntaxOn
\str_new:N \l_test_str
\str_set:Nn \l_test_str {abc}
\text_titlecase:n {\l_test_str}
\ExplSyntaxOff
\end{document}
```
Version numbers:
```
This is XeTeX, Version 3.141592653-2.6-0.999993 (TeX Live 2021) (preloaded format=xelatex)
LaTeX2e <2021-06-01> patch level 1
L3 programming layer <2021-08-27>
```

**Comment by josephwright** [comment]
Others may disagree, but to me this seems predictable: a string isn't 'text' in a TeX sense (which will contain characters of 'standard' category codes).

**Comment by josephwright** [comment]
In particular, here we are dealing with the 'first letter' rule for titlecasing: we've not found a first 'letter' so everything gets case changed.

**Comment by josephwright** [comment]
The core issue here is that to detect the first 'letter' we decided to use the catcode of each token: that seems reasonable in most cases. We could I guess cover strings by checking the catcode of the char in general too, but this seems to be getting a  bit involved.

**Comment by blefloch** [comment]
I think this is a confusing choice and I agree that `"Abc"` would be more expected, namely using ``\int_compare:nTF { 11 = \char_value_catcode:n {`#1} }`` instead of `\token_if_eq_catcode:NNTF X #1`

**Comment by josephwright** [comment]
I think I went this way as we are expecting 'text', which implies a somewhat predictable catcode regime. However, we can certainly adjust here: I guess the question is whether to have _just_ `\char_value_catcode:n` or to use a `\bool_lazy_or:nn` test and cover both?

**Comment by Skillmon** [comment]
@josephwright why not only some fast check for the catcode? Something like

```tex
\prg_new_conditional:Npnn \my_if_letter:N #1 { TF }
  {
    \exp_after:wN \__my_if_letter:w \tex_the:D \tex_catcode:D `#1 \exp_stop_f:
      \__my_if_letter_true:w 11
    \if_false:
      \prg_return_true:
    \else:
      \prg_return_false:
    \fi:
  }
\cs_new:Npn \__my_if_letter:w #1 11 {}
\cs_new:Npn \__my_if_letter_true:w 11 \if_false: { \if_true: }
```

**Comment by josephwright** [comment]
@Skillmon Speed is not the issue here: all of the token-by-token mappings are relatively slow. It's a question of what we want to support.

**Comment by josephwright** [comment]
I suspect I'll just have to make a call here ...

**Comment by josephwright** [comment]
This I suspect links to https://github.com/latex3/latex3/pull/1141 - if we take the latter approach, the same data structure could be used to save the Unicode class information into the format. I can then adjust the case changer to look at the Unicode class of character tokens rather than the catcode - that would be as expected by the Unicode algorithm and would be more appropriate in particular in pdfTeX. Thoughts?

**Comment by josephwright** [comment]
Anyone got any thoughts on my plan here? We have merged #1141 so it's now pretty trivial to save Unicode character class data and to query codepoints even in 8-bit engines.

**Comment by josephwright** [comment]
OK, coming back to this, I have Unicode class available now, and (hopefully) titlecasing tightened up - I'll make a branch to use class 'properly' here
